### ISSUE TITLE
Question: Max Input Length and Max Output Length disregarded on batch_sort_key == "shuffle"?

### ISSUE THREAD

**Issue by yash-s20** [espnet/espnet#1701]
Hi, I was interested in loading two datasets on the fly, and decided to switch the batch_sort_key to "shuffle" mode for one of them. However I notice in the source code that on shuffle, adaptive batching isn't used (adaptive batching btw, thumbs up, is a great idea). While this would otherwise be no issue, suddenly my GPUs don't have enough space to load many utterances of very long length.
How should I use a random shuffle of data with adaptive batching? 
Because if I cannot, the only option would be to reduce batch_size to 1 per GPU, and massively hit my training time.
Thanks!

**Comment by sw005320** [comment]
Actually, `batch_sort_key == 'shuffle'` is intended to disable adaptive batching with max Input length and max output length. Could you explain more about why you want to use `batch_sort_key == 'shuffle'`? Note that we perform shuffling across mini-batches, which is enough for shuffling the data. Also, shuffling utterances inside a mini-batch does not change the result (it slightly changes the result due to the numerical issue), and some functions require to sort the utterance lengths inside a mini-batch for the efficient GPU computation.

**Comment by yash-s20** [comment]
Shuffling across mini-batches is ok.
I was tweaking around to perform [mixup](https://arxiv.org/abs/1710.09412) on 2 datasets. In the case where both loaded datasets are the same, unless I use shuffle, both load the same features in the same order and mixup becomes useless. Utterance lengths are not an issue, but combining them with non-adaptive batching and the GPU cannot fit the same.

**Comment by yash-s20** [comment]
I was thinking, could I just shuffle the list of minibatches at will after collecting them in adaptive style? Would that break anything?

**Comment by sw005320** [comment]
Hmm. We may hack to do it, but it might be better to read the multiple datasets separately and modify https://github.com/espnet/espnet/blob/master/espnet/asr/pytorch_backend/asr.py#L140
This would be more flexible.

**Comment by yash-s20** [comment]
Thanks guys!
